### PR TITLE
HW4

### PR DESCRIPTION
**Pull request by DimonKhutornoy**

### CODE REVIEW

File: include/afina/concurrency/Executor.h [hunk 1]
```diff
@@ -28,16 +29,41 @@ class Executor {
...
 
     /**
      * Signal thread pool to stop, it will stop accepting new jobs and close threads just after each become
      * free. All enqueued jobs will be complete.
      *
      * In case if await flag is true, call won't return until all background jobs are done and all threads are stopped
      */
-    void Stop(bool await = false);
+    void Stop(bool await = false) {
+        std::unique_lock<std::mutex> lock(this->mutex);
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:52]: 1) Stop должно быть возможно вызывать 2 раза.
Stop(true) после Stop(false) - корректный вызов, сейчас из-за detach не будут дожидаться

2) Сейчас segfault
abort() в std\:\:thread \:\:~thread()

3) Deadlock - this->mutex требуется для завершения потока

File: include/afina/concurrency/Executor.h [hunk 2]
```diff
@@ -28,16 +29,41 @@ class Executor {
...
 
-    Executor(std::string name, int size);
-    ~Executor();
+    Executor(std::string name, int size, size_t min_threads = 2, size_t max_threads = 4, size_t wt_time = 3000)
+        : _name(std::move(name)), max_queue_size(size), low_watermark(min_threads), high_watermark(max_threads),
+          wt_time(wt_time) {
+        std::unique_lock<std::mutex> lock(this->mutex);
+        for (int i = 0; i < low_watermark; ++i) {
+            threads.emplace_back(std::thread([this]{return thread_worker(this);}));
+        }
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:40]: state - не установлен и *не инициализирован*

File: include/afina/concurrency/Executor.h [hunk 3]
```diff
@@ -71,7 +103,33 @@ class Executor {
...
+                {
+                    std::unique_lock<std::mutex> cv_lock(executor->mutex);
+                    while (executor->tasks.empty()) {
+                        executor->empty_condition.wait_for(cv_lock, std::chrono::milliseconds(executor->wt_time));
+                        if (executor->tasks.empty() || executor->threads.size() > executor->low_watermark) {
+                            auto this_thread_id = std::this_thread::get_id();
+                            for (auto it = executor->threads.begin(); it < executor->threads.end(); it++) {
+                                if (it->get_id() == this_thread_id) {
+                                    auto self = std::move(*it);
+                                    executor->threads.erase(it);
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: 1) Сначала detach
2) erase из вектора, по которому идет итерирование в Stop

File: include/afina/concurrency/Executor.h [hunk 4]
```diff
@@ -71,7 +103,33 @@ class Executor {
     /**
      * Main function that all pool threads are running. It polls internal task queue and execute tasks
      */
-    friend void perform(Executor *executor);
+    friend void thread_worker(Executor *executor)
+    {
+            executor->mutex.lock();
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: unique_lock? Потом сразу unlock?
Вынести std::unique_lock<std::mutex> cv_lock(executor->mutex) до цикла

File: include/afina/concurrency/Executor.h [hunk 5]
```diff
@@ -71,7 +103,33 @@ class Executor {
...
+            while (executor->state == Executor::State::kRun) {
+                executor->mutex.unlock();
+                std::function<void()> task;
+                {
+                    std::unique_lock<std::mutex> cv_lock(executor->mutex);
+                    while (executor->tasks.empty()) {
+                        executor->empty_condition.wait_for(cv_lock, std::chrono::milliseconds(executor->wt_time));
+                        if (executor->tasks.empty() || executor->threads.size() > executor->low_watermark) {
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: 1) Наверное, имелось в виду &&
2) Если мы проснулись при завершении сервера, то нужно исполнить все оставшиеся таски. Сейчас исполнится одна. Заменить условие на executor->state != Executor::State::kStopped
3. Нужно предусмотреть случай случайного пробуждения - проверять, истек ли таймаут. Проще использовать wait_until, чтобы не продлевать время ожидания

File: include/afina/concurrency/Executor.h [hunk 6]
```diff
@@ -51,15 +77,21 @@ class Executor {
...
 
         std::unique_lock<std::mutex> lock(this->mutex);
-        if (state != State::kRun) {
+        if (state != State::kRun || tasks.size() >= max_queue_size) {
             return false;
         }
 
         // Enqueue new task
+
+        if (threads.size() < high_watermark) {
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Нужно проверять, нет ли свободных потоков. И только если их нет, создавать новые

File: src/network/mt_blocking/ServerImpl.cpp [hunk 7]
```diff
@@ -149,8 +151,11 @@ void ServerImpl::OnRun() {
         {
             std::lock_guard<std::mutex> lock(_cs_mutex);
             if (_client_sockets.size() < _max_worker) {
-                std::thread(&ServerImpl::Worker, this, client_socket).detach();
-                _client_sockets.insert(client_socket);
+                void (ServerImpl::*func)(int);
+                func = &ServerImpl::Worker;
```
Comment (Fen99) [inline src/network/mt_blocking/ServerImpl.cpp]: Объединить с предыдущей строкой

File: src/network/mt_blocking/ServerImpl.cpp [hunk 8]
```diff
@@ -149,8 +151,11 @@ void ServerImpl::OnRun() {
         {
             std::lock_guard<std::mutex> lock(_cs_mutex);
             if (_client_sockets.size() < _max_worker) {
```
Comment (Fen99) [inline src/network/mt_blocking/ServerImpl.cpp]: За _max_workers должен следить сам пул потоков

Comment (Fen99) [reply src/network/mt_blocking/ServerImpl.cpp]: И ошибка с добавлением таски на завершении

File: src/network/mt_blocking/ServerImpl.cpp [hunk 9]
```diff
@@ -113,6 +114,7 @@ void ServerImpl::OnRun() {
     Protocol::Parser parser;
     std::string argument_for_command;
     std::unique_ptr<Execute::Command> command_to_execute;
+    Afina::Concurrency::Executor executor{"ClientSockets", 8};
```
Comment (Fen99) [inline src/network/mt_blocking/ServerImpl.cpp]: Убрать логику Join-а к воркерам из ServerImpl, за нее отвечает тредпул

**Review by Fen99 — CHANGES_REQUESTED** [review]

File: include/afina/concurrency/Executor.h [hunk 10]
```diff
@@ -28,16 +29,39 @@ class Executor {
...
+        : _name(std::move(name)), max_queue_size(size), low_watermark(min_threads), high_watermark(max_threads),
+          wt_time(wt_time) {
+        std::unique_lock<std::mutex> lock(this->mutex);
+        for (int i = 0; i < low_watermark; ++i) {
+            threads.emplace_back(std::thread([this]{return thread_worker(this);}));
+        }
+    }
+
+    ~Executor() {
+        if (state != State::kStopped) {
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Stop(true);
Не нужен if - проверка на kStopped без мьютекса, внутри Stop() - правильная проверка

File: include/afina/concurrency/Executor.h [hunk 11]
```diff
@@ -28,16 +29,39 @@ class Executor {
...
+    void Stop(bool await = false) {
+        std::unique_lock<std::mutex> lock(this->mutex);
+        state = State::kStopping;
+        empty_condition.notify_all();
+        for (auto &thread : threads) {
+            if (await) {
+                thread.join();
+            }
+        }
+        threads.clear();
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Если вызвать Stop(false), то будет terminate

File: include/afina/concurrency/Executor.h [hunk 12]
```diff
@@ -71,7 +101,34 @@ class Executor {
...
+    friend void thread_worker(Executor *executor)
+    {
+            std::unique_lock<std::mutex> cv_lock(executor->cv_mutex);
+            while (executor->state == Executor::State::kRun) {
+                std::function<void()> task;
+                {
+                    while (executor->tasks.empty()) {
+                        executor->empty_condition.wait_for(cv_lock, std::chrono::milliseconds(executor->wt_time));
+                        if (executor->tasks.empty() && executor->threads.size() > executor->low_watermark) {
+                            auto this_thread_id = std::this_thread::get_id();
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:118]: 1. Берем this->mutex
2. Не удаляем, если kStopping

File: include/afina/concurrency/Executor.h [hunk 13]
```diff
@@ -51,15 +75,21 @@ class Executor {
         auto exec = std::bind(std::forward<F>(func), std::forward<Types>(args)...);
 
         std::unique_lock<std::mutex> lock(this->mutex);
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Берем cv_mutex

File: include/afina/concurrency/Executor.h [hunk 14]
```diff
@@ -28,16 +29,39 @@ class Executor {
...
     /**
      * Signal thread pool to stop, it will stop accepting new jobs and close threads just after each become
      * free. All enqueued jobs will be complete.
      *
      * In case if await flag is true, call won't return until all background jobs are done and all threads are stopped
      */
-    void Stop(bool await = false);
+    void Stop(bool await = false) {
+        std::unique_lock<std::mutex> lock(this->mutex);
+        state = State::kStopping;
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:54]: Нужно взять cv_mutex, изменить state, потом notify, unlock

File: include/afina/concurrency/Executor.h [hunk 15]
```diff
@@ -51,15 +75,21 @@ class Executor {
...
         std::unique_lock<std::mutex> lock(this->mutex);
-        if (state != State::kRun) {
+        if (state != State::kRun || cur_queue_size >= max_queue_size) {
             return false;
         }
 
         // Enqueue new task
+
+        if (threads.size() < high_watermark && cur_queue_size++ > 0 && threads.size() == worked_threads) {
+            threads.emplace_back(std::thread([this] { return thread_worker(this); }));
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Берем mutext, если нужно запустить поток

File: include/afina/concurrency/Executor.h [hunk 16]
```diff
@@ -71,7 +101,34 @@ class Executor {
...
+    {
+            std::unique_lock<std::mutex> cv_lock(executor->cv_mutex);
+            while (executor->state == Executor::State::kRun) {
+                std::function<void()> task;
+                {
+                    while (executor->tasks.empty()) {
+                        executor->empty_condition.wait_for(cv_lock, std::chrono::milliseconds(executor->wt_time));
+                        if (executor->tasks.empty() && executor->threads.size() > executor->low_watermark) {
+                            auto this_thread_id = std::this_thread::get_id();
+                            for (auto it = executor->threads.begin(); it < executor->threads.end(); it++) {
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: std::find, он умеет работать с предикатом

File: include/afina/concurrency/Executor.h [hunk 17]
```diff
@@ -71,7 +101,34 @@ class Executor {
...
      */
-    friend void perform(Executor *executor);
+    friend void thread_worker(Executor *executor)
+    {
+            std::unique_lock<std::mutex> cv_lock(executor->cv_mutex);
+            while (executor->state == Executor::State::kRun) {
+                std::function<void()> task;
+                {
+                    while (executor->tasks.empty()) {
+                        executor->empty_condition.wait_for(cv_lock, std::chrono::milliseconds(executor->wt_time));
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: wait_until, т.к. иначе может быть долгое ожидание из-за внезапных пробуждений.
Нужно продумать 4 варианта событий:
1. Пробуждение из-за новой таски
2. Пробуждение из-за Stop - проверить, что оставшиеся таски исполнятся
3. Пробуждение из-за времени
4. Случайное пробуждение

File: include/afina/concurrency/Executor.h [hunk 18]
```diff
@@ -71,7 +101,34 @@ class Executor {
...
+                                }
+                            }
+                        }
+                    }
+                    task = std::move(executor->tasks.front());
+                    executor->tasks.pop_front();
+                    executor->cur_queue_size--;
+                    executor->worked_threads++;
+                }
+                task();
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:134]: Отпустить мьютекс на исполнение таски, потом взять снова

**Review by Fen99 — CHANGES_REQUESTED** [review]

File: include/afina/concurrency/Executor.h (hunk 1 above)
Comment (Fen99) [reply include/afina/concurrency/Executor.h:51]: std::unique_lock<std::mutex> _lock(this->cv_mutex);
if (!await && state == State::kStopping) { return; }

File: include/afina/concurrency/Executor.h [hunk 19]
```diff
@@ -50,16 +72,24 @@ class Executor {
...
 
         // Enqueue new task
+        {
+            std::unique_lock<std::mutex> lock(this->mutex);
+            if (threads.size() < high_watermark && cur_queue_size++ > 0 && threads.size() == worked_threads) {
+                threads.emplace_back(std::thread([this] { return thread_worker(this); }));
+            }
+        }
+
+        std::unique_lock<std::mutex> _lock(this->mutex);
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Убрать

File: include/afina/concurrency/Executor.h (hunk 1 above)
Comment (Fen99) [reply include/afina/concurrency/Executor.h:52]: std::unique_lockstd::mutex _lock(this->cv_mutex);
if (!await && state == State::kStopping) { return; }

File: include/afina/concurrency/Executor.h [hunk 20]
```diff
@@ -71,7 +101,39 @@ class Executor {
...
+                                std::find_if(executor->threads.begin(), executor->threads.end(),
+                                             [this_thread_id](std::thread &x) { return x.get_id() == this_thread_id; });
+                            executor->threads.erase(it);
+                            return;
+                        }
+                    }
+                }
+                task = std::move(executor->tasks.front());
+                executor->tasks.pop_front();
+                executor->cur_queue_size--;
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:129]: executor->tasks.size()

File: include/afina/concurrency/Executor.h [hunk 21]
```diff
@@ -71,7 +101,39 @@ class Executor {
...
+        auto now = std::chrono::system_clock::now();
+        while (executor->state == Executor::State::kRun) {
+            std::function<void()> task;
+            {
+                while (executor->tasks.empty()) {
+                    auto wake_stat = executor->empty_condition.wait_until(
+                        cv_lock, now + std::chrono::milliseconds(executor->wt_time));
+                    if (executor->tasks.empty() && executor->threads.size() > executor->low_watermark &&
+                        wake_stat == std::cv_status::timeout) {
+                        std::unique_lock<std::mutex> lock(executor->mutex);
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Мьютекс под условие

File: src/network/mt_blocking/ServerImpl.cpp [hunk 22]
```diff
@@ -113,6 +115,7 @@ void ServerImpl::OnRun() {
...
```
Comment (Fen99) [inline src/network/mt_blocking/ServerImpl.cpp]: Не 8, а n_workers (?)

File: include/afina/concurrency/Executor.h [hunk 23]
```diff
@@ -71,7 +101,39 @@ class Executor {
...
+    friend void thread_worker(Executor *executor) {
+        std::unique_lock<std::mutex> cv_lock(executor->cv_mutex);
+        auto now = std::chrono::system_clock::now();
+        while (executor->state == Executor::State::kRun) {
+            std::function<void()> task;
+            {
+                while (executor->tasks.empty()) {
+                    auto wake_stat = executor->empty_condition.wait_until(
+                        cv_lock, now + std::chrono::milliseconds(executor->wt_time));
+                    if (executor->tasks.empty() && executor->threads.size() > executor->low_watermark &&
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Если state != kRun, то надо завершаться, для этого return нужно тоже вынести из if-а ниже

File: include/afina/concurrency/Executor.h [hunk 24]
```diff
@@ -71,8 +100,40 @@ class Executor {
...
+                        return;
+                    }
+                }
+                task = std::move(executor->tasks.front());
+                executor->tasks.pop_front();
+                executor->cur_queue_size--;
+                executor->worked_threads++;
+            }
+            cv_lock.unlock();
+            task();
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Еще здесь нужно сделать try...catch (или в месте, где происходит bind)

**Review by Fen99 — CHANGES_REQUESTED** [review]

**Review by Fen99 — APPROVED** [review]
